src/deploy/functions/checkIam.ts

```diff
@@ -104,3 +107,82 @@ export async function checkHttpIam(
   }
   logger.debug("[functions] found setIamPolicy permission, proceeding with deploy");
 }
+
+function reduceEventsToServices(services: Array<Service>, endpoint: backend.Endpoint) {
+  const service = serviceForEndpoint(endpoint);
+  if (!services.find((s) => s.name === service.name)) {
+    services.push(service);
+  }
+  return services;
+}
+
+/**
+ * Checks and sets the roles for specific resource service agents
+ * @param projectId project identifier
+ * @param want backend that we want to deploy
+ * @param have backend that we have currently deployed
+ */
+export async function ensureServiceAgentRoles(
+  projectId: string,
+  want: backend.Backend,
+  have: backend.Backend
+): Promise<void> {
+  // find new services
+  const wantServices = backend.allEndpoints(want).reduce(reduceEventsToServices, []);
+  const haveServices = backend.allEndpoints(have).reduce(reduceEventsToServices, []);
+  const newServices = wantServices.filter(
+    (wantS) => !haveServices.find((haveS) => wantS.name === haveS.name)
+  );
+  if (newServices.length === 0) {
+    return;
+  }
+  // get the full project iam policy
+  let policy: iam.Policy;
+  try {
+    policy = await getIamPolicy(projectId);
+  } catch (err) {
+    logger.warn(
+      "We failed to obtain the IAM policy for the project,",
+      " the function deployment might fail if service ",
+      "accounts do not have the correct assigned roles."
+    );
+    return;
+  }
+  // run in parallel all the missingProjectBindings jobs
+  const findUpdatedBindings: Array<Promise<Array<iam.Binding>>> = [];
+  newServices.forEach((service) =>
+    findUpdatedBindings.push(service.missingProjectBindings(projectId, policy))
```

Comment:
thought: does it get any simpler if service just returns `requiredProjectBindings` rather than `missingProjectBindings`? We already handle the case where the policy already exists.

Reply:
It doesn't really get any simpler, just shaves off a couple lines. But since we already have the logic to merge bindings, i went ahead and changed the function to always return the required bindings.